src/Coalesce.Synchronizer/service/src/main/java/com/incadencecorp/coalesce/synchronizer/service/operations/VerifyOperationImpl.java

```diff
@@ -0,0 +1,188 @@
+
+    @Override
+    protected AbstractOperationTask createTask()
+    {
+        return new AbstractOperationTask() {
+
+            private final Set<String> missing = new HashSet<>();
+
+            @Override
+            protected Boolean doWork(String[] keys, CachedRowSet rowset) throws CoalescePersistorException
+            {
+                missing.addAll(verify(target, keys));
+
+                if (!missing.isEmpty())
+                {
+                    if (LOGGER.isDebugEnabled())
+                    {
+                        LOGGER.debug(CoalesceErrors.NOT_FOUND.replaceAll("%s", "{}"), "Entity", missing);
+                    }
+
+                    throw new CoalescePersistorException(String.format(CoalesceErrors.NOT_FOUND, "Entity", missing));
+                }
+
+                return true;
+            }
+
+            @Override
+            public String[] getErrorSubset()
+            {
+                return missing.toArray(new String[missing.size()]);
+            }
+        };
+    }
+
+    /**
+     * @param target to verify
+     * @param keys   which should be present within the target
+     * @return a list of keys that are missing from the target.
+     */
+    public static Set<String> verify(ICoalescePersistor target, String[] keys) throws CoalescePersistorException
+    {
+        Set<String> missing;
+
+        if (target instanceof ICoalesceSearchPersistor)
+        {
+            missing = verifyBySearch((ICoalesceSearchPersistor) target, keys);
+        }
+        else
+        {
+            missing = verifyByCRUD(target, keys);
+        }
+
+        return missing;
+    }
+
+    private static Set<String> verifyByCRUD(ICoalescePersistor persister, String[] keys) throws CoalescePersistorException
+    {
+        Set<String> results = new HashSet<>();
+
+        // Get from Source
+        CoalesceEntity[] entities = persister.getEntity(keys);
+
+        if (keys.length != entities.length)
+        {
+            Set<String> entityKeys = new HashSet<>();
+
+            for (CoalesceEntity entity : entities)
+            {
+                entityKeys.add(entity.getKey());
+            }
+
+            for (String key : keys)
+            {
+                if (!entityKeys.contains(key))
+                {
+                    results.add(key);
+                }
+            }
+
+        }
+
+        return results;
+    }
+
+    private static Set<String> verifyBySearch(ICoalesceSearchPersistor persister, String[] keys)
+            throws CoalescePersistorException
+    {
+        Set<String> results = new HashSet<>();
+
+        List<Filter> filters = new ArrayList<>();
+
+        for (String key : keys)
+        {
+            filters.add(CoalescePropertyFactory.getEntityKey(key));
+        }
+
+        Query query = new Query();
+        query.setMaxFeatures(keys.length);
+        query.setFilter(CoalescePropertyFactory.getFilterFactory().or(filters));
```

Comment:
Question,   what index gets searched here? Is there always a Coalesce index with all entities?

Reply:
Every database is going to have an index which contains the XML. Based on the filter is up to the persister to determine which index that is. For Elastic this would be coalesce.

Reply:
This operation verifies that the list of keys exists within the coalesce index. Where that list comes from is dependent on the scanner. You can obtain that list from another database's coalesce index or from a recordset's index. So as an example you could create a SQL statement to pull all entity keys from the coalesce-oedocument index and verify they exists within the coalesce index etc. On an error (entity key does not exists) an exception is thrown in which case the driver pulls the list of keys that had issues and handles them. By handle this is based on the configuration. If no handler is specified the exception is propagated otherwise its dependent on the handlers implementation. Using the FileHandler it will write the keys that had issues to the file system with the exception as the body.